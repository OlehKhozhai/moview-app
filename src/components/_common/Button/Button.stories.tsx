import React, { FC } from 'react';
import { Meta } from '@storybook/react';

import Button, { ButtonVariation, ButtonProps } from './index';

export default {
  component: Button,
  title: 'components/Button',
  argTypes: {
    variation: {
      options: [ButtonVariation.Primary, ButtonVariation.PrimaryOutline, ButtonVariation.Secondary],
      control: { type: 'radio' },
      defaultValue: ButtonVariation.Primary,
    },
    title: { control: 'text' },
    disabled: { control: 'boolean' },
    onClick: { control: 'action' },
  },
} as Meta;

export const Default: FC<ButtonProps> = (props) => <Button {...props}>Button title</Button>;
