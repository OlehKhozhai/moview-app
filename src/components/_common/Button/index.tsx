import React from 'react';
import { Link } from 'react-router-dom';
import cn from 'classnames';

import styles from './styles.module.scss';

type ButtonProps = {
  title?: string | React.ReactNode;
  type?: 'button' | 'submit';
  linkTo?: string;
  variation?: 'primary' | 'primaryOutline' | 'secondary';
  children?: React.ReactNode;
  className?: string;
  disabled?: boolean;
  onClick?: () => void;
};

const Button: React.FC<ButtonProps> = ({
  title,
  type = 'button',
  variation = 'primary',
  linkTo,
  children,
  className,
  disabled = false,
  onClick,
}) => {
  return linkTo ? (
    <Link to={linkTo} className={cn(styles.root, styles[variation], className)}>
      {title || children}
    </Link>
  ) : (
    <button
      className={cn(styles.root, styles[variation], className)}
      type={type}
      disabled={disabled}
      onClick={onClick}
    >
      {title || children}
    </button>
  );
};

export default React.memo(Button);
