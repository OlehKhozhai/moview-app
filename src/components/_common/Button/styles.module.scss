@import 'assets/styles/mixins/colors';

.root {
  padding: 15px 25px;
  border: none;
  font-size: 20px;
  border-radius: 5px;
  color: $primary-color;
  cursor: pointer;
  text-transform: uppercase;
  text-decoration: none;
}

.primary {
  background-color: $secondary-bg;

  &:hover {
    background-color: $secondary-dark-bg;
  }
}

.primaryOutline {
  background-color: transparent;
  border: 1px solid $primary-border;
  color: $secondary-color;
}

.secondary {
  background-color: $primary-light-bg;
  color: $secondary-color;

  &:hover {
    background-color: $primary-dark-bg;
  }
}
