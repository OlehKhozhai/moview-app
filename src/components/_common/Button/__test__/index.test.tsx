import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { create } from 'react-test-renderer';

import Button from '../index';

describe('Button component', () => {
  test('should render Button component by default', () => {
    const component = create(<Button title="Button title" />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Button component with all props', () => {
    const component = create(
      <Button title="Button title" onClick={() => undefined} type="button" variation="secondary" />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Button component with children', () => {
    const component = create(
      <Button onClick={() => undefined} type="button" variation="secondary">
        Button title
      </Button>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Button component as link with children', () => {
    const component = create(
      <MemoryRouter initialEntries={['/']}>
        <Button onClick={() => undefined} linkTo="test.com" variation="secondary">
          Button title
        </Button>
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Button component as a link', () => {
    const component = create(
      <MemoryRouter initialEntries={['/']}>
        <Button title="Button title" linkTo="test.com" />
      </MemoryRouter>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should trigger onClick method', () => {
    const handleOnClick = jest.fn();
    const component = create(<Button title="Button title" onClick={handleOnClick} />);

    component.root.props.onClick();

    expect(handleOnClick).toHaveBeenCalled();
  });
});
