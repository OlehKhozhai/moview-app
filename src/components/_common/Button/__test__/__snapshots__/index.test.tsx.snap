// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render Button component as a link 1`] = `
<a
  className="root primary"
  href="/test.com"
  onClick={[Function]}
>
  Button title
</a>
`;

exports[`Button component should render Button component as link with children 1`] = `
<a
  className="root secondary"
  href="/test.com"
  onClick={[Function]}
>
  Button title
</a>
`;

exports[`Button component should render Button component by default 1`] = `
<button
  className="root primary"
  disabled={false}
  type="button"
>
  Button title
</button>
`;

exports[`Button component should render Button component with all props 1`] = `
<button
  className="root secondary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Button title
</button>
`;

exports[`Button component should render Button component with children 1`] = `
<button
  className="root secondary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Button title
</button>
`;
